parser/encoding/request.go

```diff
@@ -0,0 +1,138 @@
+		overrideDefault: true,
+	},
+	"qs": {
+		location:        Query,
+		overrideDefault: true,
+	},
+	"header": {
+		location:        Header,
+		overrideDefault: true,
+	},
+	"json": {
+		location:        Body,
+		overrideDefault: false,
+	},
+}
+
+// RequestEncoding expresses how a request should be encoded for a specified set of HttpMethods
+type RequestEncoding struct {
+	// The HTTP methods these field configurations can be used for
+	HttpMethods []string
+	// Contains metadata about how to marshal a http parameter
+	Fields []*ParameterEncoding
+}
+
+// ParameterEncoding expresses how a parameter should be encoded
+type ParameterEncoding struct {
+	// Location (e.g. header, query, body) where the parameter should be marshalled from/to
+	Location ParamLocation
+	// The location specific name of the parameter (e.g. cheeseEater, cheese-eater, X-Cheese-Eater
+	ParamName string
+	// The underlying field definition
+	Field *v1.Field
+}
+
+// DescribeRequest combines a payload v1.Struct and a list of http methods and returns
+// a list of RPCRequest metadata. RPCRequest specifies which of the requested http method
+// it can be used for.
+func DescribeRequest(payload *v1.Struct, httpMethods ...string) ([]*RequestEncoding, error) {
+	methodsByLocation := make(map[ParamLocation][]string)
+	for _, m := range httpMethods {
+		switch m {
+		case "GET", "HEAD", "DELETE":
+			methodsByLocation[Query] = append(methodsByLocation[Query], m)
+		case "*":
+			methodsByLocation[Body] = []string{"POST", "PUT", "PATCH"}
+			methodsByLocation[Query] = []string{"GET", "HEAD", "DELETE"}
+		default:
+			methodsByLocation[Body] = append(methodsByLocation[Body], m)
+		}
+	}
+
+	var reqs []*RequestEncoding
+	for location, methods := range methodsByLocation {
+		fields, err := describeParams(location, payload)
+		if err != nil {
+			return nil, err
+		}
+		reqs = append(reqs, &RequestEncoding{
+			HttpMethods: methods,
+			Fields:      fields,
+		})
+	}
+	return reqs, nil
+}
+
+// describeParams calls describeParam() for each field in the payload struct
+func describeParams(defaultLocation ParamLocation, payload *v1.Struct) (fields []*ParameterEncoding, err error) {
+	for _, f := range payload.GetFields() {
+		f, err := describeParam(defaultLocation, f)
+		if err != nil {
+			return nil, err
+		}
+		fields = append(fields, f)
+	}
+	return fields, nil
+}
+
+// formatName formats a parameter name with the default formatting for the location (e.g. snakecase for query)
+func formatName(location ParamLocation, name string) string {
+	switch location {
+	case Query:
+		return parser.SnakeCase(name)
+	default:
+		return name
+	}
+}
+
+// describeParam returns an RPCField which falls back on defaultLocation if no parameter
+// (e.g. qs, query, header) tag is set
+func describeParam(defaultLocation ParamLocation, field *v1.Field) (*ParameterEncoding, error) {
+	rpcField := &ParameterEncoding{defaultLocation, formatName(defaultLocation, field.Name), field}
+	var usedTag string
```

Comment:
Maybe rename to be clearer you didn't intend to set it outside of the overrideDefault?

```suggestion
	var usedOverrideTag string
```